src/seat/input_method_v3.rs

```diff
@@ -0,0 +1,711 @@
+    /// Bind the input_method global, if it exists
+    pub fn bind<D>(globals: &GlobalList, qh: &QueueHandle<D>) -> Result<Self, BindError>
+    where
+        D: Dispatch<XxInputMethodManagerV2, GlobalData> + 'static,
+    {
+        let manager = globals.bind(qh, 2..=2, GlobalData)?;
+        Ok(Self { manager })
+    }
+
+    /// Request a new input_method object associated with a given
+    /// seat.
+    pub fn get_input_method<State>(&self, qh: &QueueHandle<State>, seat: &WlSeat) -> InputMethod
+    where
+        State: Dispatch<XxInputMethodV1, InputMethodData, State> + 'static,
+    {
+        InputMethod {
+            input_method: self.manager.get_input_method(
+                seat,
+                qh,
+                InputMethodData::new(seat.clone()),
+            ),
+        }
+    }
+
+    pub fn get_positioner<State>(&self, qh: &QueueHandle<State>) -> PopupPositioner
+    where
+        State: Dispatch<XxInputPopupPositionerV1, PositionerData, State> + 'static,
+    {
+        PopupPositioner(self.manager.get_positioner(qh, PositionerData))
+    }
+}
+
+impl<D> Dispatch<xx_input_method_manager_v2::XxInputMethodManagerV2, GlobalData, D>
+    for InputMethodManager
+where
+    D: Dispatch<xx_input_method_manager_v2::XxInputMethodManagerV2, GlobalData>
+        + InputMethodHandler,
+{
+    fn event(
+        _data: &mut D,
+        _manager: &xx_input_method_manager_v2::XxInputMethodManagerV2,
+        _event: xx_input_method_manager_v2::Event,
+        _: &GlobalData,
+        _conn: &Connection,
+        _qh: &QueueHandle<D>,
+    ) {
+        unreachable!()
+    }
+}
+
+/// A trivial wrapper for an [`XxInputPopupPositionerV1`].
+///
+/// This wrapper calls [`destroy`][XxInputPopupPositionerV1::destroy] on the contained
+/// positioner when it is dropped.
+#[derive(Debug)]
+pub struct PopupPositioner(XxInputPopupPositionerV1);
+
+impl Deref for PopupPositioner {
+    type Target = XxInputPopupPositionerV1;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Drop for PopupPositioner {
+    fn drop(&mut self) {
+        self.0.destroy()
+    }
+}
+
+impl<D> Dispatch<XxInputPopupPositionerV1, PositionerData, D> for PopupPositioner
+where
+    D: Dispatch<XxInputPopupPositionerV1, PositionerData> + InputMethodHandler,
+{
+    fn event(
+        _data: &mut D,
+        _manager: &XxInputPopupPositionerV1,
+        _event: xx_input_popup_positioner_v1::Event,
+        _: &PositionerData,
+        _conn: &Connection,
+        _qh: &QueueHandle<D>,
+    ) {
+        unreachable!("Positioner has no events")
+    }
+}
+
+#[derive(Debug)]
+pub struct PositionerData;
+
+#[derive(Debug)]
+pub struct InputMethod {
+    input_method: XxInputMethodV1,
+}
+
+impl InputMethod {
+    pub fn input_method(&self) -> &XxInputMethodV1 {
+        &self.input_method
+    }
+
+    pub fn set_preedit_string(&self, text: String, cursor: CursorPosition) {
+        // TODO: should this enforce indices on codepoint boundaries?
```

Comment:
I think that would be good. It could be changed to return a Result, with an error when there is an index not on a boundary?

Reply:
done.